interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala

```diff
@@ -157,100 +162,110 @@ private class CatsEffect
-  override final def handleErrorWith[A](fa: IO[E, A])(f: E => IO[E, A]): IO[E, A] = fa.catchAll(f)
-  override final def raiseError[A](e: E): IO[E, A]                                = IO.fail(e)
+private class CatsMonadError[R, E] extends CatsMonad[R, E] with MonadError[ZIO[R, E, ?], E] {
+  override final def handleErrorWith[A](fa: ZIO[R, E, A])(f: E => ZIO[R, E, A]): ZIO[R, E, A] = fa.catchAll(f)
+  override final def raiseError[A](e: E): ZIO[R, E, A]                                        = ZIO.fail(e)
 }
 
 /** lossy, throws away errors using the "first success" interpretation of SemigroupK */
-private trait CatsSemigroupK[E] extends SemigroupK[IO[E, ?]] {
-  override final def combineK[A](a: IO[E, A], b: IO[E, A]): IO[E, A] = a.orElse(b)
+private trait CatsSemigroupK[R, E] extends SemigroupK[ZIO[R, E, ?]] {
+  override final def combineK[A](a: ZIO[R, E, A], b: ZIO[R, E, A]): ZIO[R, E, A] = a.orElse(b)
 }
 
-private class CatsAlternative[E: Monoid] extends CatsMonadError[E] with Alternative[IO[E, ?]] {
-  override final def combineK[A](a: IO[E, A], b: IO[E, A]): IO[E, A] =
+private class CatsAlternative[R, E: Monoid] extends CatsMonadError[R, E] with Alternative[ZIO[R, E, ?]] {
+  override final def combineK[A](a: ZIO[R, E, A], b: ZIO[R, E, A]): ZIO[R, E, A] =
     a.catchAll { e1 =>
       b.catchAll { e2 =>
-        IO.fail(Monoid[E].combine(e1, e2))
+        ZIO.fail(Monoid[E].combine(e1, e2))
       }
     }
-  override final def empty[A]: IO[E, A] = raiseError(Monoid[E].empty)
+  override final def empty[A]: ZIO[R, E, A] = raiseError(Monoid[E].empty)
 }
 
-trait CatsBifunctor extends Bifunctor[IO] {
-  override final def bimap[A, B, C, D](fab: IO[A, B])(f: A => C, g: B => D): IO[C, D] =
+trait CatsBifunctor[R] extends Bifunctor[ZIO[R, ?, ?]] {
+  override final def bimap[A, B, C, D](fab: ZIO[R, A, B])(f: A => C, g: B => D): ZIO[R, C, D] =
     fab.bimap(f, g)
 }
 
-private class CatsParallel[E](final override val monad: Monad[IO[E, ?]]) extends Parallel[IO[E, ?], ParIO[E, ?]] {
+private class CatsParallel[R, E](final override val monad: Monad[ZIO[R, E, ?]])
+    extends Parallel[ZIO[R, E, ?], ParIO[E, ?]] {
 
   final override val applicative: Applicative[ParIO[E, ?]] =
     new CatsParApplicative[E]
 
-  final override val sequential: ParIO[E, ?] ~> IO[E, ?] =
-    new (ParIO[E, ?] ~> IO[E, ?]) { def apply[A](fa: ParIO[E, A]): IO[E, A] = Par.unwrap(fa) }
+  final override val sequential: ParIO[E, ?] ~> ZIO[R, E, ?] =
+    new (ParIO[E, ?] ~> ZIO[R, E, ?]) { def apply[A](fa: ParIO[E, A]): ZIO[R, E, A] = Par.unwrap(fa) }
 
-  final override val parallel: IO[E, ?] ~> ParIO[E, ?] =
-    new (IO[E, ?] ~> ParIO[E, ?]) { def apply[A](fa: IO[E, A]): ParIO[E, A] = Par(fa) }
+  // FIXME: asInstanceOf should not be needed
+  final override val parallel: ZIO[R, E, ?] ~> ParIO[E, ?] =
+    new (ZIO[R, E, ?] ~> ParIO[E, ?]) {
+      def apply[A](fa: ZIO[R, E, A]): ParIO[E, A] = Par(fa.provide(().asInstanceOf[R]))
```

Comment:
Maybe we have to change the `ParIO` type synonym to make this one work without `asInstanceOf`?

Reply:
Definitely, I forgot about this one! :smile: